### ISSUE TITLE
[signal] Question about filter sweeps

### ISSUE THREAD

**Issue by harold** [generateme/fastmath#9]
Hi! Thanks for this library, it has lots of great stuff in it and I'm enjoying playing around.

I have some existing code that generates audio waveforms as sequences of samples, so the design of the effect nodes as closures that operate sequentially on samples is great.

I am able to make a `:simple-lowpass` effect and apply it to my sounds and it works as expected.

What I'd like to be able to do is vary (sweep) the cutoff frequency over time, and I don't see a way to do that currently.

https://github.com/generateme/fastmath/blob/90cda5fba2e29fc66ee06d8924ca16f0101ff1ca/src/fastmath/signal.clj#L310

Maybe it would be good if `(fastmath.signal/effect ...)` returned something that could update its parameters - I have in mind something akin to a VST effect, so parameters could be changed "during" processing.

Regardless, thank you again for this library, many useful functions here and I look forward to exploring it more.

Take care.

**Comment by genmeblog** [comment]
Hi! Thanks for using a `fastmath`! Yeah, it's quite big bag of functions :)

I don't see an easy way to make this working as you described. Currently effect keeps only internal state not the configuration and allows to reset state.
The true is I created that for an image manipulation techniques (sonification for glitch art) rather than real DSP :)

If you wish we can try to refactor it and to enable more vst or ladspa way ie: input/output channels and state management. I believe it's doable.

**Comment by harold** [comment]
> I don't see an easy way to make this working as you described.
>

No worries! I saw the same, thank you for the confirmation.

> The true is I created that for an image manipulation techniques (sonification for glitch art) rather than real DSP :)
>

Sounds fun! That makes sense.

> If you wish we can try to refactor it and to enable more vst or ladspa way ie: input/output channels and state management. I believe it's doable.
>

I also believe it's doable, though it might require changes that touch a lot of the code in `signal.clj`.

This is not critical for me at the moment. I have other directions that I am going to take this project in first. I am grateful to you for this project and this signal code in particular. I am learning a lot by reading and using it.

Take care!

**Comment by genmeblog** [comment]
Since you can build your own effects you can do the following:

```clojure
(defn sweep-alpha
  [^double rate]
  (let [tinterval (/ rate)
        t (atom (long 0))] ;; to know the sample id
    (fn []
      (let [cutoff (m/norm (m/cos (* 100.0 tinterval (swap! t clojure.core/inc))) 1.0 -1.0 500.0 5000.0)
            tau (/ (* cutoff m/TWO_PI))]
        (/ tinterval (+ tau tinterval))))))

(repeatedly 10 (sweep-alpha 44100.0))
;; => (0.06650128426969248
;;     0.06650343885351115
;;     0.06650702979213088
;;     0.06651205703393591
;;     0.06651852050666654
;;     0.06652642011742149
;;     0.06653575575266107
;;     0.06654652727821102
;;     0.06655873453926664
;;     0.06657237736039835)

(defmethod effect :simple-lowpass-sweep
  ([m] (effect m {}))
  ([m {:keys [rate]
       :or {rate 44100.0}}]
   (let [next-alpha (sweep-alpha rate)]
     (effect-node m (fn
                      ([^double sample ^double prev]
                       (let [alpha (next-alpha)
                             s1 (* sample alpha)
                             s2 (- prev (* prev alpha))
                             nprev (+ s1 s2)]
                         (SampleAndState. nprev nprev)))
                      ([] 0.0))))))
```

`cutoff` is generated from sine and (in above example)  bounces between 500.0 to 5000.0Hz 100 times per second.

**Comment by harold** [comment]
Ha! That's great.

[square-lowpass-sweep.wav.gz](https://github.com/generateme/fastmath/files/6386122/square-lowpass-sweep.wav.gz)

I am off and running.

One thing I encountered is that `fastmath.signal/effect-node` is a private function. This means that to define my effect in my own file I ended up writing this (note the `#'signal/effect-node` bit):

```
(defmethod signal/effect :simple-lowpass-sweep
  ([m] (signal/effect m {}))
  ([m {:keys [rate]
       :or {rate 44100.0}}]
   (let [alpha-fn (produce-alpha-fn rate)]
     (#'signal/effect-node
      m (fn
          ([^double sample ^double prev]
           (let [alpha (alpha-fn)
                 s1 (* sample alpha)
                 s2 (- prev (* prev alpha))
                 nprev (+ s1 s2)]
             (SampleAndState. nprev nprev)))
          ([] 0.0))))))
```

**Comment by genmeblog** [comment]
Oh, that's cool! Glad it works. Actually this filter is really simple with quite flat slope. Anyway, it works :)
I can add state variable filter, which I have implemented elsewhere (https://github.com/genmeblog/soundsynth)
I will correct the visibility of crucial functions and publish a snapshot soon.

**Comment by harold** [comment]
Yup! A multipole filter (steeper slope) and/or one with resonance would be appreciated.

It appears we have overlapping interests. 😄

**Comment by genmeblog** [comment]
Yeah :) However writing the software synthesizer was kind of a study to help with my son's project. 

Eventually I've made `effect-node` as a public function. Sorry for delay. 
No SVF yet. I have to figure out how to make parameters in terms of frequencies.

**Comment by harold** [comment]
Perhaps you son and I are the ones with overlapping interests then, ha!

Thanks for your thoughts here. I have finished the project I was working on in this area for now, but when I come back to it I will integrate fastmath deeper.

Take care.
